lib/Dialect/TritonGPU/Transforms/RemoveLayoutConversions.cpp

```diff
@@ -82,542 +82,676 @@ class DecomposeDotOperand : public mlir::RewritePattern {
-        if (cvts.contains(&op)) {
-          if (mapping.contains(op.getOperand(0)))
-            mapping.map(op.getResult(0), mapping.lookup(op.getOperand(0)));
-          continue;
-        }
-        cloneWithInferType(rewriter, &op, mapping);
+SmallVector<Value> LayoutPropagation::propagateToUsers(Value value,
+                                                       LayoutInfo &info) {
+  SmallVector<Value> changed;
+  for (OpOperand &use : value.getUses()) {
+    Operation *user = use.getOwner();
+    if (auto forOp = dyn_cast<scf::ForOp>(user)) {
+      Value arg = forOp.getRegionIterArgForOpOperand(use);
+      Value result = forOp.getResultForOpOperand(use);
+      setEncoding({arg, result}, info, changed, user);
+      continue;
+    }
+    if (auto yieldOp = dyn_cast<scf::YieldOp>(user)) {
+      auto parent = yieldOp->getParentOp();
+      SmallVector<Value> valuesToPropagate = {
+          parent->getResult(use.getOperandNumber())};
+      if (auto forOp = dyn_cast<scf::ForOp>(parent)) {
+        valuesToPropagate.push_back(
+            forOp.getRegionIterArg(use.getOperandNumber()));
+        setEncoding({valuesToPropagate}, info, changed, user);
       }
-    };
-    rewriter.setInsertionPointToEnd(newIfOp.thenBlock());
-    rematerialize(ifOp.thenBlock(), thenCvts);
-    if (hasElse) {
-      rewriter.setInsertionPointToEnd(newIfOp.elseBlock());
-      rematerialize(ifOp.elseBlock(), elseCvts);
+      // TODO: handle scf.if and while.
+      continue;
     }
+    // Workaround: don't propagate through truncI
+    if (isa<arith::TruncIOp>(user))
+      continue;
+    if (user->hasTrait<mlir::OpTrait::SameOperandsAndResultEncoding>() ||
+        user->hasTrait<mlir::OpTrait::Elementwise>() ||
+        isa<triton::ReduceOp, triton::ExpandDimsOp,
+            triton::gpu::ConvertLayoutOp>(user)) {
+      setEncoding(user->getResults(), info, changed, user);
+      continue;
+    }
+  }
+  return changed;
+}
 
-    rewriter.setInsertionPointAfter(newIfOp);
-    SmallVector<Value> newRetValues = newIfOp.getResults();
-    for (size_t i = 0; i < numOps; i++) {
-      if (newIfOp.getResult(i).getType() != ifOp.getResult(i).getType()) {
-        newRetValues[i] = rewriter.create<triton::gpu::ConvertLayoutOp>(
-            newIfOp.getLoc(), ifOp.getResult(i).getType(),
-            newIfOp.getResult(i));
+void LayoutPropagation::propagateLayout() {
+  SmallVector<Value> queue;
+  for (auto it : layouts) {
+    queue.push_back(it.first);
+  }
+  while (!queue.empty()) {
+    Value currentValue = queue.back();
+    LayoutInfo &info = layouts[currentValue];
+    queue.pop_back();
+    SmallVector<Value> changed = propagateToUsers(currentValue, info);
+    queue.insert(queue.end(), changed.begin(), changed.end());
+  }
+}
+
+void LayoutPropagation::resolveConflicts() {
+  for (auto &it : layouts) {
+    LayoutInfo &info = it.second;
+    if (info.encodings.size() <= 1)
+      continue;
+    // Hacky resolve, prefer block encoding.
```

Comment:
What does it mean prefer blocking encoding? Can you give me an example?

Reply:
it means if we have a case like:
```
%l = tt.load ... tensor<.. , #blocked>
%d = tt.dot .. -> tensor<.., #mma>
%cl = convert %l
%dl = convert %d
%a = arith.add %cl, %dl
```
Because the add depends on both the load with blocked layout and the dot with mma layout we need to decide between those two layouts. In this case we would pick the blocked layout and there would be a convert between the dot and the add. This is a bit arbitrary as building a clever heuristic would require a more complex cost model and this allows us to not change the layout picked for flash attention compare to existing code.